### DISCUSSION TITLE
Sluggish on windows

### DISCUSSION THREAD

**Discussion by AmerM137** [ibhagwan/fzf-lua#1597] [General]
Hello,

I didn't want to post an issue as this might be specific to windows and not exactly a bug. I am testing fzf-lua today with a very simple config (mini.deps as plugin manager, treesitter, basic lspconfig, conform). I've used telescope before and most recently mini.pick before and they are both fairly fast, but it seems that fzf-lua is "sluggish" in comparison.

### Setup: 
fast workstation i9-11900k w/ nvme drive
fzf can index my home directory in just a few seconds (~1.2 million files)
windows 11
neovim 0.10.2
fzf 0.57.0
fd/rg/bat available path
I am testing this in a very small directory, basically my %LOCALAPPDATA%\nvim directory. Only a handful of files present.
All relevant folders (scoop/, nvim/, nvim-data/) are excluded from windows defender.

### Description
By sluggish that I mean I've observed the following things, **mostly UI related, not search related**:

- The entering and exiting of the UI is a little sluggish. There is a good 200ms+ of latency when invoking the command and exiting after picking a file.
-  The preview for default profile is a little laggy, even on smallish files, especially when flipping through files quickly.
- After selecting a file to open, it takes ~200-300ms to open the fille, even if the file is already in the buffer list, or if it's even the same file. My nvim startup is way faster than that, even on cold boot.
- I tried the "telescope" and "max-perf" profiles. The telescope profile improves the previewer but not the launching/exiting of the UI. The max-perf profile's previewer is slower than telescope, all else feels the same.

@ibhagwan what else should I try to increase performance on windows? Is there anything else I may have missed? I'd love to use this plugin as I use fzf on the cli with powershell all the time and it's very fast, but it seems that inside neovim it's sluggish in comparison.

**Comment by AmerM137** [comment]
I guess I am not the only one with this issue, I just read the other posts about the latency on windows. Sorry for duplicating.

**Comment by ibhagwan** [comment]
Unforuntaely I'm not sure what I can do about this, Windows has a very inefficient way of managing processes (compared to Unix based OS's) and there's not much I can do about this,  `fzf.vim` behaves execatly the same, that's the nature of having to open a new process for every operation, Telescope on the otherhand doesn't need to spawn the fzf process and run a terminal it just runs `rg/fd` in the background and displays the results so it's spwaning and terminating will always be faster.

The benefit of using an external process is disconnect from neovim's main thread which results in lag free and great performance when searching / fuzzy matching but that does come at a cost on Windows.

Some users reported significant improvement after excluding the neovim/fzf/rg/fd in Windows defender, you can try that.

Other than that my only advice would be to use WSL which IMHO works super fast as if it was Linux.

**Comment by ibhagwan** [comment]
A little more specific:

> The entering and exiting of the UI is a little sluggish. There is a good 200ms+ of latency when invoking the command and exiting after picking a file.

My notes regarding windows process spawn apply

> The preview for default profile is a little laggy, even on smallish files, especially when flipping through files quickly.

For a previe to trigger, fzf needs to call an fzf lua neovim headless wrapper with some data (callback function ID + the items), again this involves a process spawn, same issue.

> After selecting a file to open, it takes ~200-300ms to open the fille, even if the file is already in the buffer list, or if it's even the same file. My nvim startup is way faster than that, even on cold boot.

Perhaps related to process termination? I can't really tell my Windows VPS was a "donation" and was very slow, I couldn't never attempt to optimize it due to how painfully slow it was, I couldn't tell wether it was the very low-end spec VPS or another issue, all I could observe was that `fzf.vim` had the same issue and that WSL2 was significantly faster.

> I tried the "telescope" and "max-perf" profiles. The telescope profile improves the previewer but not the launching/exiting of the UI. The max-perf profile's previewer is slower than telescope, all else feels the same.

`max-perf` uses the `bat` previewer which is yet another process spawn (although in the shell, owned by fzf) so again, the same mechanic applies.

Try to exclude all these binaries from Windows defender, should help I hope.

**Comment by AmerM137** [comment]
Binaries and nvim folders are excluded already. Thanks for the detail! I guess windows will just be windows.

**Reply by ibhagwan** [reply]
> Binaries and nvim folders are excluded already. Thanks for the detail! I guess windows will just be windows.

Sucks indeed.
